hudi-common/src/main/java/org/apache/hudi/common/table/TableSchemaResolver.java

```diff
@@ -257,6 +259,8 @@ private Option<Schema> getTableSchemaFromCommitMetadata(HoodieInstant instant, b
       Schema schema = new Schema.Parser().parse(existingSchemaStr);
       if (includeMetadataFields) {
         schema = HoodieAvroUtils.addMetadataFields(schema, hasOperationField.get());
+      } else {
+        schema = HoodieAvroUtils.removeMetadataFields(schema);
```

Comment:
In which case, the schema contains the metadata fields and we have to remove explicitly?

Reply:
After alter table schema like 
`alter table t1 drop f_date`, `alter table t1 add f_long bigint`, the schema of alter schema instant would contain metadata fields.
If not remove those metadata fields, the insert query would throw following exception
![Uploading image.png…]()

Reply:
> the schema of alter schema instant would contain metadata fields.

Should we fix that?

Reply:
cc @xiarixiaoyao The schema of alter schema instant including metadata fields, is this allowable?

Reply:
BTW, I think it's reasonable to remove metadata fields if `includeMetadataFields ` is false.

Reply:
agree with remove metadata fields if includeMetadataFields is false.

Only internalSchema requires metadata fields,

Spark on hudi handles metadata fields on its own, so whether there are metadata fields in the avor schema does not have much impact on Spark's read and write operations. 

However, from the perspective of function definition, it is indeed necessary to remove metadata fields here